rayon-core/src/registry.rs

```diff
@@ -499,6 +503,19 @@ unsafe fn main_loop(worker: Worker<JobRef>, registry: Arc<Registry>, index: usiz
     // **user code** panics, we should catch that and redirect.
     let abort_guard = unwind::AbortIfPanic;
 
+    // Inform a user callback that we started a thread.
+    if let Some(ref handler) = registry.start_handler {
+        let registry = registry.clone();
+        match unwind::halt_unwinding(|| handler()) {
+            Ok(()) => {
+            }
+            Err(err) => {
+                registry.handle_panic(err);
+            }
+        }
+        registry.terminate()
```

Comment:
OK, calling `registry.terminate()` in *this* location seems definitely wrong, since it will execute unconditionally. I could *imagine* that you might want to call it in the case of panic, but then it should move within the arm. I'm not entirely sure however if invoking `terminate` here will work; I'd have to review the logic.